---
title: "CentralDogmaGroup7"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CentralDogmaGroup7}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Link to repository

<https://github.com/rforbiodatascience24/group_7_package>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(CentralDogmaGroup7)
```

### Function 1

Function 1 is a random DNA sequence generator. It takes an input size for the DNA sequence length, then makes a vector of the given length with any combination of "A", "T", "G", and "C" with replacement. Then it concatenates the vector to a string with no spaces between the letters and returns the concatenated string.

### Function 2

The `transcription_rna()` function is part of the central dogma simulation, designed to take a vector of DNA codons and convert it into their RNA form. It returns the complete RNA sequence.

### Function 3

Function three, `codon_extractor()` converts a nucleic acid sequence into it's respective codons. It takes a string input (e.g. 'GCCATATAG') and returns a new string containing the codons of the input sequence (e.g. 'GCC ATA TAG').

### Function 4

The `translation()` function is part of the central dogma simulation, designed to take a vector of RNA codons and convert it into a protein sequence. It maps each codon to its corresponding amino acid using a codon table, concatenates the amino acids, and returns the complete protein sequencet.

### Function 5

The function `plot_occurrence()` is development around the central dogma, created with the intention of taking in an amino acid sequence, and outputting a plot of the occurrence of each amino acid.

It takes in a character string, splits it into individual characters, and creates a new string containing only the unique characters. Then, it counts the occurrence of each unique character, creating a data frame with the number of times each character appears. Finally, it plots the occurrence of each unique character in a bar plot.

## Used in conjunction

Together, the above function simulate the central dogma from creating a DNA sequence, through transcription and translation, and analyzing the amino acid rate of occurrence.

1.  A DNA sequence of the length 10.000 is created:

```{r}
#| output: false

DNA <- random_DNA(10000)
```

2.  The DNA is transcribed into the coding strand of it's counterpart mRNA:

```{r}
#| output: false

mRNA <- transcription(DNA)
```

3.  The mRNA sequence is divided into triplets:

```{r}
#| output: false

codons <- codon_extractor(mRNA)
```

4.  The triplets are translated into a sequence of amino acids:

```{r}
#| output: false

amino_acids <- translation(codons)
```

5.  The amino acids are quantified and their occurrence is plotted:

```{r}
plot_occurrence(amino_acids)
```

## Uses and considerations

-   Used to simulate central dogma, eventually creating a visual output of amino acid occurrence.

-   Can be tweaked to general handling of strings and occurrences.

-   Dependencies, as the word implies, makes the package dependent on the quality of other packages.

-   A bug in a dependency, can cause a bug in the package.

-   More dependencies increase the size and load time.

-   Time and consideration has been put in to specialized packages, such as ggplot - redoing functions from this package would cause unnecessary waste of time and energy, developing and troublehooting already existing advanced functions.

-   Compatibility with other standards and packages, such as tidyverse, is ensured by utilizing already existing packages.

-   `@importFrom package function` imports a specific function from a package, allowing us to call the function without specifying the function name

    -   can be an advantage when function is used often

    -   makes the code appear cleaner and more readable

-   `package::function()` calls the function from the package as it is being used

    -   makes it clear from which package it being called

    -   keeps the namespace clean
