% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={CentralDogmaGroup07},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}

\title{CentralDogmaGroup07}
\author{}
\date{\vspace{-2.5em}}

\begin{document}
\maketitle

\subsection{Link to repository}\label{link-to-repository}

\url{https://github.com/rforbiodatascience24/group_7_package}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{library}\NormalTok{(CentralDogmaGroup7)}
\end{Highlighting}
\end{Shaded}

\subsubsection{Function 1}\label{function-1}

Function 1 is a random DNA sequence generator. It takes an input size
for the DNA sequence length, then makes a vector of the given length
with any combination of ``A'', ``T'', ``G'', and ``C'' with replacement.
Then it concatenates the vector to a string with no spaces between the
letters and returns the concatenated string.

\subsubsection{Function 2}\label{function-2}

The \texttt{transcription\_rna()} function is part of the central dogma
simulation, designed to take a vector of DNA codons and convert it into
their RNA form. It returns the complete RNA sequence.

\subsubsection{Function 3}\label{function-3}

Function three, \texttt{codon\_extractor()} converts a nucleic acid
sequence into it's respective codons. It takes a string input
(e.g.~`GCCATATAG') and returns a new string containing the codons of the
input sequence (e.g.~`GCC ATA TAG').

\subsubsection{Function 4}\label{function-4}

The \texttt{translation()} function is part of the central dogma
simulation, designed to take a vector of RNA codons and convert it into
a protein sequence. It maps each codon to its corresponding amino acid
using a codon table, concatenates the amino acids, and returns the
complete protein sequence.

\subsubsection{Function 5}\label{function-5}

The function \texttt{plot\_occurrence()} is development around the
central dogma, created with the intention of taking in an amino acid
sequence, and outputting a plot of the occurrence of each amino acid.

It takes in a character string, splits it into individual characters,
and creates a new string containing only the unique characters. Then, it
counts the occurrence of each unique character, creating a data frame
with the number of times each character appears. Finally, it plots the
occurrence of each unique character in a bar plot.

\subsection{Used in conjunction}\label{used-in-conjunction}

Together, the above function simulate the central dogma from creating a
DNA sequence, through transcription and translation, and analyzing the
amino acid rate of occurrence.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  A DNA sequence of the length 10.000 is created:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{DNA }\OtherTok{\textless{}{-}} \FunctionTok{random\_DNA}\NormalTok{(}\DecValTok{10000}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{1}
\tightlist
\item
  The DNA is transcribed into the coding strand of it's counterpart
  mRNA:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mRNA }\OtherTok{\textless{}{-}} \FunctionTok{transcription}\NormalTok{(DNA)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{2}
\tightlist
\item
  The mRNA sequence is divided into triplets:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{codons }\OtherTok{\textless{}{-}} \FunctionTok{codon\_extractor}\NormalTok{(mRNA)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{3}
\tightlist
\item
  The triplets are translated into a sequence of amino acids:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{amino\_acids }\OtherTok{\textless{}{-}} \FunctionTok{translation}\NormalTok{(codons)}
\end{Highlighting}
\end{Shaded}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\setcounter{enumi}{4}
\tightlist
\item
  The amino acids are quantified and their occurrence is plotted:
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{plot\_occurrence}\NormalTok{(amino\_acids)}
\end{Highlighting}
\end{Shaded}

\includegraphics{CentralDogmaGroup7_files/figure-latex/unnamed-chunk-96-1.pdf}

\subsection{Uses and considerations}\label{uses-and-considerations}

\subsubsection{The central dogma}\label{the-central-dogma}

The code's use for simulating the central dogma, is limited. The
opportunity to print the string after each individual step of the
biological process, makes it a nice tool for understanding the basic
principals. However, running the functions in succession multiple time,
reveals a undesirable issue: the frequent occurrence of stop codons. In
nature, these are sparsely distributed, but our \texttt{random\_DNA()}
function, takes no such thing into account. This leads to a high chance
of translations being interrupted. If the initial function 1 is not run,
however, the remaining functions of the code is capable of quantifying
and visualizing the abundance of each amino acid while simulating the
central dogma.

\subsubsection{Development}\label{development}

The issue mentioned aboved, could be fixed by altering function 1 to
take into account the natural occurrence of stop codons. This alteration
would include utilizing real data to quantify the abundance of specific
codons, and could be added to make the code better to simulate real life
events.

Another addition, if we want our code to better simulate reality, would
be to include the concept of terminators for the transcription. Such
addition would entail identifying palindromes, by tracking the
characters of the DNA string and identifying inverted repeats.

Many more such functions could be added, with the help of research on
the subject, expanding the package to be able to identify sites of
interest on both the DNA and mRNA sequence, eventually creating a
package capable of more advanced bioinformatics.

\subsubsection{Dependencies}\label{dependencies}

Dependencies, as the word implies, makes the package dependent on the
quality of other packages. There are both pros and cons to this. E.g. a
bug in a dependency, can cause a bug in the package, and more
dependencies increase the size and load time. On the other hand, time
and consideration has been put into specialized packages, such as
ggplot, and redoing functions from this package would cause unnecessary
waste of time. Often, the functions we wish to use already exists in
other packages. Furthermore, compatibility with other standards and
packages, such as tidyverse, is ensured by utilizing already existing
packages.

Applying the above mentioned ideas, would probably be a lot faster using
already existing functions from previously developed packages, allowing
us to focus on collecting necessary functions for the specific purpose
of our package.

When using dependencies, it is important to consider the difference
between \texttt{@importFrom\ package\ function} and
\texttt{package::function()}. \texttt{@importFrom\ package\ function}
imports a specific function from a package, allowing us to call the
function without specifying the function name. It can be an advantage
when a function is used often and makes the code appear cleaner and more
readable. \texttt{package::function()} calls the function from the
package as it is being used. This makes it clear in the script from
which package it being called and keeps the namespace clean, allowing
for easier editing of the package later on.

\end{document}
